New Documentation - Introduction - General Purpose Backend (#1889)
kuzzleio/kuzzle: doc/2/guides/introduction/2-general-purpose-backend/index.md

---
code: false
type: page
title: General Purpose Backend
description: Kuzzle main concepts and features overview
order: 200
---

# General Purpose Backend

Kuzzle is a generic backend offering features common to all applications.

> Rather than redeveloping functionalities that don't bring any added value, Kuzzle proposes to use them as is to build its high-level business functionalities.

Author: I rewrited it

Rather than redeveloping functionalities that **do not bring value for your users**, Kuzzle proposes to use its basic functionalities in order to **build high-level business functionalities**.

> Rather than redeveloping functionalities that don't bring any added value, Kuzzle proposes to use them as is to build its high-level business functionalities.

Reviewer: ```suggestion
Kuzzle ships with the basic building blocks needed by any application out-of-the box, letting you focus on features that really add value to your product.
```

## Extensibility

Like any framework, Kuzzle allows you to develop new features by extending and modifying the existing API.

To do so, just install the NPM kuzzle package and start developing your application.

```js
import { Backend, Request } from 'kuzzle'

const app = new Backend('iot-tracker')

// Register a new API controller
app.controller.register('greeting', {
  actions: {
    sayHello: {
      handler: (request: request) => `Hello, ${request.input.args.name}`
    }
  }
})

app.start()
  .then(() => app.log.info('Application started'))
```

Kuzzle offers different mechanisms to develop the business functionalities of its application:
 - API Controller
 - Event System
 - External Plugin

All the developed functionalities will benefit from the advantages of the Kuzzle API such as multi-authentication, rights management, standard request and response format, cluster scalability, anti-dos protection, etc.

## API First

> The majority of Kuzzle's functionalities are available via its API for various external clients.

Reviewer: ```suggestion
The majority of Kuzzle's features are available via its API for various external clients.
```

This multi-protocol API allows clients to communicate with Kuzzle and use the backend features through the protocol that best suits their needs.

Whether it is the creation and modification of the database or the management of users and rights, everything is available through the different controllers of the API.

The API provides a standard communication format for requests and responses so that every client has the same experience.

Access to each action is thus centralized within the rights system for a better understanding and maintenance.

## Unleash the Frontend

With Kuzzle, customers can directly access all the data stored in the database as long as their have the rights to do so.

> No more need to create a new controller every time new data needs to be displayed, no more need to add parameters to controller actions to refine searches: everything is done directly on the client side.

Reviewer: ```suggestion
No more need to create a new controller every time new data needs to be displayed, no more need to add parameters to controller actions to refine searches: queries are expressed directly on the client side.
```

---
code: false
type: page
title: General Purpose Backend
description: Kuzzle main concepts and features overview
order: 200
---

# General Purpose Backend

Kuzzle is a generic backend offering **the basic building blocks common to every application**.

> Rather than redeveloping functionalities that **do not bring value for your users**, Kuzzle proposes to use its basic building blocs in order to **build high-level business features**.

Reviewer 2: ```suggestion
Rather than redeveloping functionalities that **do not bring value for your users**, Kuzzle proposes to use its basic building blocks in order to **build high-level business features**.
```

## API First

The majority of Kuzzle's features are available via its [API](/core/2/guides/main-concepts/1-api) for various external clients.

This **multi-protocol API** allows clients to communicate with Kuzzle and use the backend features through the **protocol that best suits their needs**.

Whether it is the creation and modification of the database or the management of users and rights, **everything is available through the different controllers of the API**.

The API provides a **standard communication format** for requests and responses so that every client has the same experience.

Access to each action is thus **centralized within the rights management system** for a better understanding and maintenance.

## Ready-to-use Database

> Kuzzle use Elasticsearch as a [NoSQL document store](/core/2/guides/main-concepts/2-data-storage).

Reviewer 2: ```suggestion
Kuzzle uses Elasticsearch as a [NoSQL document store](/core/2/guides/main-concepts/2-data-storage).
```

> With Kuzzle, customers **can directly access all the data stored in the database** as long as their have the rights to do so.

Reviewer 2: ```suggestion
With Kuzzle, customers **can directly access data stored in the database** as long as they have the rights to do so.
```

> No more need to create a new controller every time new data needs to be displayed, no more need to add parameters to controller actions to refine searches: **queries are expressed directly on the client side**.

Reviewer 2: ```suggestion
It's no longer needed to create a new controller every time new data need to be displayed, and it's no longer needed to add parameters to controller actions to refine searches either: **queries are expressed directly on the client's side**.
```

In the same way, creating collections or the writing of documents is also done directly from the frontend into the database collections:

```js
// First create an index and a collection to handle our data
await sdk.index.create('iot')
await sdk.collection.create('iot', 'sensors', {
  mappings: {
    model: { type: 'keyword' },
    temperature: { type: 'integer' },
  }
})

// Create a document inside our collection
let result = await sdk.document.create('iot', 'sensors', {
  model: 'temperature',
  temperature: 42
})
```

## Authentication

Kuzzle features a [multi-authentication system](/core/2/guides/main-concepts/5-authentication) for users.

Rather than using a single authentication system, Kuzzle embeds [Passport.js](http://www.passportjs.org/) and **makes available the use of its 500+ authentication strategies** by writing an authentication plugin.

So your users can choose to authenticate themselves with a classic **login / password** but also from an **OAuth** provider such as Facebook or Google, an **LDAP directory**, via **SAML**, etc.

Kuzzle fits perfectly in a **context of SSO and centralization of authentication** within an information system.

## Right Management

> [Rights Management](/core/2/guides/main-concepts/4-permissions) is already integrated in the backend functionalities. The **rights can be configured through Kuzzle API** or through our Admin Console.

Reviewer 2: ```suggestion
[Rights Management](/core/2/guides/main-concepts/4-permissions) is already integrated in the backend functionalities. The **rights can be configured through the Kuzzle API** or through our Admin Console.
```

Kuzzle has a **standard system with 3 dimensions**. Roles control access to API actions, profiles are a composition of roles and finally users are a composition of profiles.

[Users, Profiles and Roles](./profiles-roles.png)

> This system allows to manage 90% of the access control rights situations. For the remaining 10% Kuzzle **allows to dynamically restrict access rights** via its event system and pipe mechanism.

Reviewer 2: ```suggestion
This system allows to manage the majority of access control rights situations. For the most advanced cases, Kuzzle **allows to dynamically restrict access rights** via its event system and its pipe mechanism.
```

  for (const document of documents) {
    if (request.context.user._id !== document._source._kuzzle_info.creator) {
      throw new ForbiddenError(`Not allowed to access document ${document._id}`)
    }
  }

  return documents
})
```

## Extensibility

Like any framework, Kuzzle allows you to **develop new features by extending and modifying the existing API**.

To do so, just install the [NPM kuzzle package](https://www.npmjs.com/package/kuzzle) and start developing your application.

```js
import { Backend, Request } from 'kuzzle'

const app = new Backend('iot-tracker')

// Register a new API controller
app.controller.register('greeting', {
  actions: {
    sayHello: {
      handler: (request: request) => `Hello, ${request.input.args.name}`
    }
  }
})

app.start()
  .then(() => app.log.info('Application started'))
```

Kuzzle offers different mechanisms to **develop the business functionalities** of its application:
 - [API Controller](/core/2/guides/develop-on-kuzzle/2-api-controllers)
 - [Event System](/core/2/guides/develop-on-kuzzle/3-event-system)
 - [External Plugin](/core/2/guides/develop-on-kuzzle/4-external-plugins)

> Everything you will build upon your application will **benefit from the advantages of Kuzzle API** such as multi-authentication, rights management, standard request and response format, cluster scalability, anti-dos protection, etc.

Reviewer 2: ```suggestion
Everything you will build upon your application will **benefit from the advantages of Kuzzle API** such as multi-authentication, rights management, standard request and response format, cluster scalability, anti-DoS protection, etc.
```

## Embrace Typescript

> Kuzzle exposes [interfaces written in Typescript](/core/2/framework/classes) for a **faster learning curve** and a **better maintainability** of the applications.

Reviewer 2: ```suggestion
Kuzzle exposes [interfaces written in Typescript](/core/2/framework/classes) for a **faster learning curve** and a **better application maintainability**.
```

This support is available for writing backend applications as well as frontend applications.

Whether in the backend or in the frontend, **developers use the same interface to interact with Kuzzle**: [Kuzzle Javascript SDK](/sdk/js/7/)

The Javascript SDK is of course used by frontend applications but also in the backend, so developers can capitalize on their experience to **allow isomorphic backend / frontend development**.

## Realtime Notifications

Kuzzle has its own [high-performance realtime engine](/core/2/guides/main-concepts/6-realtime-engine).  

> This engine allows you to make pub/sub in a conventional way but also [realtime database notification](/core/2/guides/main-concepts/6-realtime-engine#database-notifications).  

Reviewer 2: ```suggestion
This engine allows you to use pub/sub communications in a conventional way, but it's also capable of triggering [realtime database notification](/core/2/guides/main-concepts/6-realtime-engine#database-notifications).  
```

> Every **change that occur on the database can generate realtime notifications**. Clients can listen database changes and **synchronize frontend or other backend application** accordingly.

Reviewer 2: ```suggestion
Every **change occuring on the database can generate realtime notifications**. Clients can listen to database changes and **synchronize frontends or other backend applications** accordingly.
```

# General Purpose Backend

Kuzzle is a generic backend offering **features common to every application**.

> Rather than redeveloping functionalities that don't bring any added value, Kuzzle proposes to use them as is to **build high-level business functionalities**.

Reviewer: I find this sentence really weird...

Author: wdyt if I add `basics`
> Rather than redeveloping `basics` functionalities that don't bring any added value, Kuzzle proposes to use them as is to **build high-level business functionalities**.

Reviewer 2: Hm... And what about this proposal?
(ping @xbill82 )

> Rather than developing the same standard features over and over again each time you create a new application, Kuzzle proposes them off-the-shelf, ready to be used, and allowing you to focus on **building high-level, high-value business functionalities**.

Author: I really like this one
